lib/utils/listener.go

```diff
@@ -37,3 +39,32 @@ func GetListenerFile(listener net.Listener) (*os.File, error) {
 	}
 	return nil, trace.BadParameter("unsupported listener: %T", listener)
 }
+
+// SingleUseListener wraps a single [net.Conn] and returns it from Accept()
+// once.
+type SingleUseListener struct {
+	conn     net.Conn
+	accepted atomic.Bool
+}
+
+// NewSingleUseListener creates a new SingleUseListener.
+func NewSingleUseListener(c net.Conn) *SingleUseListener {
+	return &SingleUseListener{
+		conn: c,
+	}
+}
+
+func (l *SingleUseListener) Accept() (net.Conn, error) {
+	if l.accepted.Swap(true) {
+		return nil, io.EOF
+	}
+	return l.conn, nil
+}
+
+func (l *SingleUseListener) Addr() net.Addr {
+	return l.conn.LocalAddr()
+}
+
+func (l *SingleUseListener) Close() error {
+	return nil
```

Comment:
What closes the underlying `Conn`? The presence of a close method in the API suggests that this will actually do something, so I'd be afraid that this could lead to leaks.

Can we improve the godoc to make it clear that single use listener does not assume ownership of conn and that it expects the caller to close the original conn?